Tasks/Maven/maventask.ts

```diff
@@ -303,70 +333,110 @@ function cleanDirectory(targetDirectory:string):Q.Promise<void> {
+        }
+        if (moduleAnalysis.analysisResults[toolName].htmlFilePath) {
+            filesToUpload.push(moduleAnalysis.analysisResults[toolName].htmlFilePath);
+        }
     }
-    if (analysisResult.htmlFilePath) {
-        filesToUpload.push(analysisResult.htmlFilePath);
+
+    if (filesToUpload.length < 1) {
+        console.log('No artifacts to upload for ' + toolName + ' analysis of module ' + moduleAnalysis.moduleName);
+        return;
     }
 
     // Copy files to a staging folder and upload the entire folder
     // This is a workaround until the following bug is fixed: https://github.com/Microsoft/vso-agent/issues/263
     filesToUpload.forEach(fileToUpload => {
-        var stagingFilePath = path.join(toolStagingDir, path.basename(fileToUpload));
+        var stagingFilePath = path.join(localStagingDir, path.basename(fileToUpload));
         tl.debug('Staging ' + fileToUpload + ' to ' + stagingFilePath);
         // Execute the copy operation. -f overwrites if there is already a file at the destination.
         tl.cp('-f', fileToUpload, stagingFilePath);
     });
 
-    console.log('Uploading artifacts for ' + analysisResult.toolName + ' from ' + toolStagingDir);
+    console.log('Uploading artifacts for ' + toolName + ' from ' + localStagingDir);
 
     // Begin artifact upload - this is an asynchronous operation and will finish in the future
     tl.command("artifact.upload", {
-        containerfolder: 'codeAnalysis',
+        // Put the artifacts in subdirectories corresponding to their module
+        containerfolder: moduleAnalysis.moduleName,
         // Prefix the build number onto the upload for convenience
-        artifactname: buildNumber + '_' + analysisResult.toolName
-    }, toolStagingDir);
+        artifactname: buildNumber + '_' + moduleAnalysis.moduleName + '_' + toolName
+    }, localStagingDir);
 }
 
 // Extract data from code analysis output files and upload results to build server
 function uploadCodeAnalysisResults():void {
     // Need to run this if any one of the code analysis tools were run
-    if (isCodeAnalysisUploadRequired()) {
-        var analysisResults:ar.AnalysisResult[] = [];
-
-        // PMD
-        if (tl.getInput('pmdAnalysisEnabled', true) == 'true') {
-            try {
-                var pmdResults:ar.AnalysisResult = pmd.processPmdOutput(sourcesDir);
-                if (pmdResults) {
-                    analysisResults.push(pmdResults);
+    var enabledCodeAnalysisTools = getEnabledCodeAnalysisTools();
+    if (enabledCodeAnalysisTools.length > 0) {
+        // Discover maven modules
+        var modules:ma.ModuleAnalysis[] = findModules(sourcesDir);
+
+        // Special case: if the root turns up as a module, the automatic name won't do
+        modules.forEach((module) => {
+            if (module.rootFolder == sourcesDir) {
+                module.moduleName = 'root'
+            }
+        });
```

Comment:
The special case code should only be needed for single-module Maven projects. 

Multiple-module maven projects will have a pom.xml in the root, but should not have any compiled code associated with the master POM, and thus should have no /target/ folder.

Said special case code will protect even in the case of multiple-module maven projects with a module at the root.